File: tests/test_bad.py
import numpy as np
import pandas as pd
from coxdev import CoxDeviance

try:
    import rpy2.robjects as rpy
    has_rpy2 = True

except ImportError:
    has_rpy2 = False

if has_rpy2:
    from rpy2.robjects.packages import importr
    from rpy2.robjects import numpy2ri
    from rpy2.robjects import default_converter

    np_cv_rules = default_converter + numpy2ri.converter

    glmnetR = importr('glmnet')
    baseR = importr('base')
    survivalR = importr('survival')

def get_glmnet_result(event,
                      status,
                      start,
                      eta,
                      weight,
                      time=False):

    event = np.asarray(event)
    status = np.asarray(status)
    weight = np.asarray(weight)
    eta = np.asarray(eta)

    with np_cv_rules.context():

        rpy.r.assign('status', status)
        rpy.r.assign('event', event)
        rpy.r.assign('eta', eta)
        rpy.r.assign('weight', weight)
        rpy.r('eta = as.numeric(eta)')
        rpy.r('weight = as.numeric(weight)')

        if start is not None:
            start = np.asarray(start)
            rpy.r.assign('start', start)
            rpy.r('y = Surv(start, event, status)')
            rpy.r('D_R = glmnet:::coxnet.deviance3(pred=eta, y=y, weight=weight, std.weights=FALSE)')
            rpy.r('G_R = glmnet:::coxgrad3(eta, y, weight, std.weights=FALSE, diag.hessian=TRUE)')
            rpy.r("H_R = attr(G_R, 'diag_hessian')")
        else:
            rpy.r('y = Surv(event, status)')
            rpy.r('D_R = glmnet:::coxnet.deviance2(pred=eta, y=y, weight=weight, std.weights=FALSE)')
            rpy.r('G_R = glmnet:::coxgrad2(eta, y, weight, std.weights=FALSE, diag.hessian=TRUE)')
            rpy.r("H_R = attr(G_R, 'diag_hessian')")

        D_R = rpy.r('D_R')
        G_R = rpy.r('G_R')
        H_R = rpy.r('H_R')

    # -2 for deviance instead of loglik

    return D_R, -2 * G_R, -2 * H_R

def get_coxph(event,
              status,
              X,
              beta,
              sample_weight,
              start=None,
              ties='efron'):

    if start is not None:
        start = np.asarray(start)
    status = np.asarray(status)
    event = np.asarray(event)

    with np_cv_rules.context():
        rpy.r.assign('status', status)
        rpy.r.assign('event', event)
        rpy.r.assign('X', X)
        rpy.r.assign('beta', beta)
        rpy.r.assign('ties', ties)
        rpy.r.assign('sample_weight', sample_weight)
        rpy.r('sample_weight = as.numeric(sample_weight)')
        if start is not None:
            rpy.r.assign('start', start)
            rpy.r('y = Surv(start, event, status)')
        else:
            rpy.r('y = Surv(event, status)')
        rpy.r('F = coxph(y ~ X, init=beta, weights=sample_weight, control=coxph.control(iter.max=0), ties=ties, robust=FALSE)')
        rpy.r('score = colSums(coxph.detail(F)$scor)')
        G = rpy.r('score')
        D = rpy.r('F$loglik')
        cov = rpy.r('vcov(F)')
    return -2 * G, -2 * D, cov

def check_results(fname, ties):
    tol = 1e-10
    rpy.r('d <- readRDS("./g0.RDS")')
    event = np.asarray(rpy.r('d$event'))
    start = None
    status = np.asarray(rpy.r('d$status'))
    weight = np.asarray(rpy.r('d$sample_weight'))
    tie_breaking = ties
    beta = np.asarray(rpy.r('d$beta'))
    X = np.array(rpy.r('d$X'))
    coxdev = CoxDeviance(event=event,
                         start=start,
                         status=status,
                         tie_breaking=tie_breaking)

    C = coxdev(X @ beta, weight)

    eta = X @ beta
    
    H = coxdev.information(eta,
                           weight)
    I = X.T @ (H @ X)
    assert np.allclose(I, I.T)
    cov_ = np.linalg.inv(I)
    
    (G_coxph,
     D_coxph,
     cov_coxph) = get_coxph(event=event,
                            status=status,
                            beta=beta,
                            sample_weight=weight,
                            start=start,
                            ties=tie_breaking,
                            X=X)
    if np.allclose(D_coxph[0], C.deviance - 2 * C.loglik_sat):
        print("Coxph Deviance matches")
    else:
        print("Coxph Deviance mismatch")
    if np.linalg.norm(G_coxph - X.T @ C.gradient) / np.linalg.norm(X.T @ C.gradient) < tol:
        print("Coxph gradient matches")
    else:
        print("Coxph gradient mismatch")
    if np.linalg.norm(cov_ - cov_coxph) / np.linalg.norm(cov_) < tol:
        print("Coxph cov matches")
    else:
        print("Coxph cov mismatch")

    if ties == 'breslow':
        D_R, G_R, H_R = get_glmnet_result(event,
                                          status,
                                          start,
                                          X @ beta,
                                          weight)
        
        delta_D = np.fabs(D_R - C.deviance) / np.fabs(D_R)
        if delta_D < tol:
            print("Glmnet Deviance matches")
        else:
            print("Glmnet Deviance mismatch")

        delta_G = np.linalg.norm(G_R - C.gradient) / np.linalg.norm(G_R)            
        if delta_G < tol:
            print("Glmnet gradient matches")
        else:
            print("Glmnet gradient mismatch")

        delta_H = np.linalg.norm(H_R - C.diag_hessian) / np.linalg.norm(H_R)
        if delta_H < tol:
            print("Glmnet hessian matches")
        else:
            print("Glmnet hessian mismatch")

check_results("g0.RDS", 'efron')
check_results("g0.RDS", 'breslow')

File: tests/test_bad.R
library(coxdev)
library(survival)
tol  <- 1e-10

all_close  <- function(a, b, rtol = 1e-05, atol = 1e-08) {
  all(abs(a - b) <= (atol + rtol * abs(b)))
}

rel_diff_norm  <- function(a, b) { a <- as.matrix(a); b <- as.matrix(b); norm(a - b, 'F') / norm(b, 'F') }

get_glmnet_result <- function(event,
                              status,
                              start,
                              eta,
                              weight) {
  event <- as.numeric(event)
  status <- as.numeric(status)
  weight <- as.numeric(weight)
  eta <- as.numeric(eta)
  if (length(start) == length(status)) {
    y <- survival::Surv(start, event, status)
    D_R <- glmnet:::coxnet.deviance3(pred=eta, y=y, weight=weight, std.weights=FALSE)
    G_R <- glmnet:::coxgrad3(eta, y, weight, std.weights=FALSE, diag.hessian=TRUE)
    H_R <- attr(G_R, 'diag_hessian')
  } else {
    y <- Surv(event, status)
    D_R <- glmnet:::coxnet.deviance2(pred=eta, y=y, weight=weight, std.weights=FALSE)
    G_R <- glmnet:::coxgrad2(eta, y, weight, std.weights=FALSE, diag.hessian=TRUE)
    H_R <- attr(G_R, 'diag_hessian')
  }
  list(D = D_R, G = as.numeric(-2 * G_R), H = -2 * H_R)
}

get_coxph <- function(event,
                      status,
                      X,
                      beta,
                      sample_weight,
                      start=None,
                      ties='efron') {

  if (length(start) == length(status)) start  <- as.numeric(start)
  start <- as.numeric(start)
  status <- as.numeric(status)
  event <- as.numeric(event)
  weight <- as.numeric(sample_weight)
  if (length(start) == length(status)) {
    y <- Surv(start, event, status)
  } else {
    y <- Surv(event, status)
  }
  F <- coxph(y ~ X, init=beta, weights=sample_weight, control=coxph.control(iter.max=0), ties=ties, robust=FALSE)
  G <- colSums(coxph.detail(F)$scor)
  D <- F$loglik
  cov <- vcov(F)
  list(G = -2 * G, D = -2 * D, cov = cov)
}

check_results  <- function(fname, ties = c('efron', 'breslow')) {
  d  <- readRDS(fname)
  ties  <- match.arg(ties)

  weight <- d$sample_weight
  event <- d$event
  start <- d$start
  status <- d$status
  cox_deviance  <- make_cox_deviance(event = event,
                                     start = start,
                                     status = status,
                                     weight = weight,
                                     tie_breaking = ties)

  X  <- d$X; tX  <- t(X);
  beta  <- d$beta
  C <- cox_deviance$coxdev(X %*% beta, weight)
  h <- cox_deviance$information(X %*% beta, weight)
  I <- tX %*% h(X)
  message(ifelse(all_close(I, t(I)), "I is symmetric enough", "I is not symmetric"))
  new_cov  <- solve(I)

  ## Check against coxph with
  coxph_result  <- get_coxph(event = event,
                             status = status,
                             beta  = beta,
                             sample_weight = weight,
                             start = start,
                             ties = ties,
                             X = X)
  G_coxph <- coxph_result$G
  D_coxph <- coxph_result$D[1]
  cov_coxph  <- coxph_result$cov

  message(ifelse(all_close(D_coxph, C$deviance - 2 * C$loglik_sat), "Coxph Deviance matches", "Coxph Deviance mismatch"))
  message(ifelse(rel_diff_norm(G_coxph, tX %*% C$gradient) < tol, "Coxph Gradient matches", "Coxph Gradient mismatch"))
  message(ifelse(rel_diff_norm(new_cov, cov_coxph) < tol, "Coxph Cov matches", "Coxph Cov mismatch"))

  if (ties == 'breslow') {
  ## Check against glmnet also
    glmnet_result  <- get_glmnet_result(event = event,
                                        status = status,
                                        start = NA,
                                        eta = X %*% beta,
                                        weight = weight)
    G_glmnet <- glmnet_result$G
    D_glmnet <- glmnet_result$D[1]
    H_glmnet  <- glmnet_result$H

    message(ifelse(all_close(D_glmnet, C$deviance), "Glmnet Deviance matches", "Glmnet Deviance mismatch"))
    message(ifelse(rel_diff_norm(G_glmnet, C$gradient) < tol, "Glmnet Gradient matches", "Glmnet Gradient mismatch"))
    message(ifelse(rel_diff_norm(H_glmnet, C$diag_hessian) < tol, "Glmnet hessian matches", "Glmnet hessian mismatch"))
  }
}

check_results("g0.RDS")

check_results("g0.RDS", 'breslow')

